### PR TITLE
feat(examples): change angular example to ts_project

### PR DESCRIPTION
**Pull request by jbedard**
@alexeagle, not sure who else was interested

So this got more complicated and took a year... :|
* prerendering was dropped because I don't have time to figure it out
* esbuild is used to merge spec files before running karma on them. This replaces concatjs, we can't simply use the compiled-ts and dynamically load all the files at runtime because those files might have cjs/amd etc.
* random other minor changes that were for some reason exposed, see comments on the diff

~~Mainly just a ~~copy~~ refactor of the angular example switching to `ts_project` with `@angular/compiler-cli` as the tsc. However running into a few issues...~~

* ~~ngc is not working with a path error:
```Error: Internal Error: absoluteFrom(src/app/home): path is not absolute"```~~ https://github.com/angular/angular/pull/36291 fixes this
* ~~trying to get a package/module_name referencing a `ts_project` and imported into another
```in deps attribute of ts_project rule //src/app/todos:todos: '//src/shared/material:material' does not have mandatory providers: 'DeclarationInfo'```~~

### CODE REVIEW

File: examples/angular_ngc/tools/angular_ts_project.bzl [hunk 1]
```diff
@@ -0,0 +1,12 @@
+"Shows how to enable both worker mode and use_angular_plugin to make a drop-in replacement for ng_module"
+
+load("@npm//@bazel/typescript:index.bzl", "ts_project")
+
+def ng_ts_project(name, tsconfig = "//:tsconfig.json", srcs = [], angular_assets = [], **kwargs):
+    ts_project(
+        name = name,
+        tsconfig = tsconfig,
+        tsc = "@npm//@angular/compiler-cli/bin:ngc",
```
Comment (alexeagle) [inline examples/angular_ngc/tools/angular_ts_project.bzl:9]: I think it should use ngtsc?

File: examples/angular_ngc/tools/angular_ts_project.bzl [hunk 2]
```diff
@@ -0,0 +1,12 @@
+"Shows how to enable both worker mode and use_angular_plugin to make a drop-in replacement for ng_module"
+
+load("@npm//@bazel/typescript:index.bzl", "ts_project")
+
+def ng_ts_project(name, tsconfig = "//:tsconfig.json", srcs = [], angular_assets = [], **kwargs):
+    ts_project(
```
Comment (alexeagle) [inline examples/angular_ngc/tools/angular_ts_project.bzl:6]: you probably need declaration = True so that we know to produce .d.ts outputs

**Review by alexeagle — APPROVED** [review]

**Comment by alexeagle** [comment]
hey @jbedard any update on getting the fix from @Toxicable on the pathing issue in the angular compiler?

**Comment by jbedard** [comment]
@alexeagle the patch from @Toxicable seems to work, I've added a patch for that to this PR. I'm still getting the `js_library` `DeclarationInfo` issue though which I don't understand... any ideas there?

**Comment by alexeagle** [comment]
I'll fetch this and repro locally

**Review by alexeagle — COMMENTED** [review]
Can you move the files so this is an edit to the existing examples/angular (so that when it lands, we use `ts_project` instead of `ts_library` for it)

otherwise this appears as 73 new files and I can't tell what you changed. I don't want to increase from 3 to 4 angular examples, it's too confusing for users and too much CI load to test them

**Comment by alexeagle** [comment]
okay so for your DeclarationInfo problem, let's diagnose starting from the target that's supposed to produce .d.ts

```
% bazel build src/shared/material:material-ng --output_groups=types
INFO: Invocation ID: 45b78a82-ddd9-4e99-aef1-a0e58ac3fd3e
INFO: Analyzed target //src/shared/material:material-ng (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //src/shared/material:material-ng up-to-date:
  dist/bin/src/shared/material/material.module.d.ts
  dist/bin/src/shared/material/material.module.d.ts.map
```

cool, there is .d.ts coming from that ng_ts_project rule. Let's see if it gets propagated through the js_library rule

```
% bazel build src/shared/material:material --output_groups=types
INFO: Invocation ID: 804d5698-39c1-488b-b34b-ff1efe93426d
INFO: Analyzed target //src/shared/material:material (80 packages loaded, 13410 targets configured).
INFO: Found 1 target...
Target //src/shared/material:material up-to-date:
  dist/bin/src/shared/material/material.module.d.ts
  dist/bin/src/shared/material/material.module.d.ts.map
```

BUT I had to make two fixes to get this result.
1) Internal: js_library forgot to propagate the OutputGroupsInfo so you couldn't ask it for "types", will fix
2) js_library propagates typings from `deps`, not `srcs`. changing your js_library to

```
js_library(
    name = "material",
    package_name = "@bazel/shorten",
    deps = [":material-ng"],
)
```
fixes it.

File: packages/typescript/internal/ts_project.bzl [hunk 3]
```diff
@@ -48,13 +48,19 @@ def _join(*elements):
 def _ts_project_impl(ctx):
     arguments = ctx.actions.args()
 
-    generated_srcs = False
+    has_generated = False
```
Comment (alexeagle) [inline packages/typescript/internal/ts_project.bzl:51]: ooh, this looks like a nice bugfix, if helpful I can repro in a standalone test case for ts_project and land as a prefactoring

Comment (jbedard) [reply packages/typescript/internal/ts_project.bzl:51]: There's 2 things here:
* if there are no source files after 1+ generated files in `ctx.files.srcs` then the `fail` would never be hit, this was a bug
* the `--rootDir` was only absolute if using generated sources. I honestly don't recall why I did this or if it correct, but I assume I did it try to to fix/workaround https://github.com/angular/angular/issues/36290 - but maybe that is only an angular bug and not ts_project bug?

If you'd like to take this commit and add tests then go for it :+1:

**Comment by jbedard** [comment]
I'm currently getting a couple errors:

```
error TS6059: File '/private/var/tmp/_bazel_jason.bedard/02195f90740dce00c87d7ba79d4dda35/sandbox/darwin-sandbox/40/execroot/examples_angular/src/app/todos/todos.module.ts' is not under 'rootDir' 'bazel-out/darwin-fastbuild/bin/src/app/todos'. 'rootDir' is expected to contain all source files.
```

```
ERROR: /tmp/tmp-16352IrbYU6I3u1uq/src/app/todos/BUILD.bazel:11:1: in ts_project rule //src/app/todos:todos:
Traceback (most recent call last):
	File "/tmp/tmp-16352IrbYU6I3u1uq/src/app/todos/BUILD.bazel", line 11
		ts_project(name = 'todos')
	File "/var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/7c6ac0f00953126366d099f5c23b87a0/external/npm/@bazel/typescript/internal/ts_project.bzl", line 61, in _ts_project_impl
		fail(<1 more arguments>)
srcs cannot be a mix of generated files and source files: bazel-out/k8-fastbuild/bin/src/app/todos/todos.component.css.map, src/app/todos/todos.component.html
```

File: packages/typescript/internal/ts_project.bzl (hunk 3 above)
Comment (jbedard) [reply packages/typescript/internal/ts_project.bzl]: I've updated that commit to output a better error messages showing the last files found in the generated vs src folders, for example the error this PR is currently getting which I commented about below:
```
srcs cannot be a mix of generated files and source files: bazel-out/k8-fastbuild/bin/src/app/todos/todos.component.css.map, src/app/todos/todos.component.html
```

**Comment by phelanma** [comment]
Hi, little question do you have a plan for this PR in the short term? (week~2 weeks period)

Just a "need to know", if I should continue to use `ts_library` over `ts_project`, as this(cli issue + internal fixes) seems like a block on the feature?

**Comment by jbedard** [comment]
This is blocked atm and needs work I think both in angular and rules_nodejs, so for now I think you'll need to stick with `ts_library`. I'd just recommend putting the angular config into a macro so you can easily change it in the future, see the macro in this PR for an example.

**Comment by literalpie** [comment]
Has there been any progress on this? Are there specific issues for what needs to be changed in rules_nodejs that I can watch?

**Comment by jbedard** [comment]
@literalpie I think this was blocked on https://github.com/angular/angular/issues/36290 but I should really rebase and give this another shot...

File: examples/angular/tools/karma.conf.js [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+  } catch (e) {
+    console.error(`Failed to extract ${archiveFile}`);
+    throw e;
+  }
+}
+
+/**
+ * Check if Chrome sandboxing is supported on the current platform.
+ */
+function supportChromeSandboxing() {
```
Comment (jbedard) [inline examples/angular/tools/karma.conf.js:115]: Should we just delete all the complications here to make the example simpler?

**Comment by literalpie** [comment]
Are there any updates on this? Are there new issues blocking it?

**Comment by alexeagle** [comment]
It's blocked on getting the Pre-rendering to work.

**Comment by jbedard** [comment]
@literalpie I have got this working in my own project using essentially the same macros as in the PR. So for simpler cases (no pre-rendering) it shouldn't be hard to setup.

**Comment by alexeagle** [comment]
@jbedard maybe we just have to remove the pre-rendering for now? I'm not sure which is worse, staying on the older rule or losing the feature.

File: examples/angular/src/BUILD.bazel [hunk 5]
```diff
@@ -176,67 +159,27 @@ terser_minified(
     src = ":bundle-es5",
 )
 
-# take the skeleton index.html and use it to render the root, /hello and /todos routes
-# this generates multiple index.html files, one for each requested route
-ng_prerender(
```
Comment (jbedard) [inline examples/angular/src/BUILD.bazel:181]: Prerendering has been dropped in hopes of getting this merged...

File: examples/angular/src/BUILD.bazel [hunk 6]
```diff
@@ -268,8 +209,8 @@ pkg_web(
 ngsw_config(
     name = "pwa",
     src = ":prodapp",
+    out = "pwa-config",
```
Comment (jbedard) [inline examples/angular/src/BUILD.bazel:212]: This just avoids a "rule and output have same name" (or something like that) warning

File: examples/angular/tools/ngsw_config.bzl [hunk 7]
```diff
@@ -11,27 +11,25 @@ def ngsw_config(name, config, index_html, src, out = None, **kwargs):
 
     nodejs_binary(
         name = ngsw_config_name,
-        data = ["@npm//@angular/service-worker", index_html, config, src],
```
Comment (jbedard) [inline examples/angular/tools/ngsw_config.bzl:14]: I think including these here is one thing that was causing an issue, although I don't fully understand why... including the `src` here (output of `ts_project`) was somehow causing the typescript compiler to re-run but without the correct tsconfig that adds the `paths`. So typescript was unnecessary re-compiling with the wrong tsconfig which failed with path errors.

File: examples/angular/tools/ngsw_config.bzl [hunk 8]
```diff
@@ -11,27 +11,25 @@ def ngsw_config(name, config, index_html, src, out = None, **kwargs):
...
+        data = ["@npm//@angular/service-worker"],
         visibility = ["//visibility:private"],
         entry_point = "@npm//:node_modules/@angular/service-worker/ngsw-config.js",
     )
 
     cmd = """
 mkdir -p $@
 cp -R $(locations {TMPL_src})/. $@/
-cp $(location {TMPL_index}) $@/index.html
```
Comment (jbedard) [inline examples/angular/tools/ngsw_config.bzl:22]: This index file is already in `TMPL_src` so this was causing a "cannot overwrite readonly file" error on ubuntu. I believe it's the exact same file so this was unnecessary (or maybe pre-rendering previously required it 🤷).

**Review by alexeagle — APPROVED** [review]
woohoo! it's green

**Pull request merged**